src/megatron/energon/task_encoder/base.py

```diff
@@ -207,19 +338,51 @@ def _batch(
         else:
             raise ValueError("Unrecognized result type.")
 
+    def pre_pack(self, samples: List[T_sample]) -> List[List[T_sample]]:
+        """
+        For packing, selects the samples to be packed together. By default, no packing is happening.
+
+        Args:
+            samples: The samples to pre-pack. A full buffer will be passed into the function.
+
+        Returns: The pre-packed samples as a list of lists of samples.
+        """
+        raise NotImplementedError("Packing only effective when overridden.")
+
+    def final_pack(self, samples: List[T_sample]) -> T_sample:
+        """
+        Given one set of samples to pack, returns the final packed sample. By default, no packing is happening.
+
+        Args:
+            samples: The samples to pack into a single sample
+
+        Returns: The final packed sample.
+        """
+        raise NotImplementedError("Packing only effective when overridden.")
+
     def build_batch(
         self,
         dataset: SavableDataset[T_encoded_sample],
         batch_size: Optional[int],
         batch_drop_last: bool = False,
+        packing_buffer_size: Optional[int] = None,
         worker_config: Optional[WorkerConfig] = None,
     ) -> SavableDataset[T_raw_batch]:
         """Applies the batcher to the dataset."""
+
+        pre_pack_provided = getattr(self.pre_pack, "__func__", None) is not TaskEncoder.pre_pack
+        final_pack_provided = (
+            getattr(self.final_pack, "__func__", None) is not TaskEncoder.final_pack
+        )
+
         if (
             getattr(self.batch_group_criterion, "__func__", None)
             is not TaskEncoder.batch_group_criterion
         ):
             assert batch_size is not None, "batch_size must be set if batch_group_criterion is set"
+            assert (
+                not pre_pack_provided and not final_pack_provided
+            ), "Packing not supported when grouping"
```

Comment:
also assert that `packing_buffer_size is None`

Reply:
Yeah good point. And I also want to allow the user to disable packing by setting `packing_buffer_size` to `None` even when the two methods are defined in the TaskEncoder. 
Tuomas I think asked how he can disable packing.